Add Request exceptions total to default metrics

deepbrook opened the pull request:
Counts exceptions arriving at ``teardown_request``.

It currently does not allow tracking further http codes in range of 405-5xx

I looked at adding the counter in the decorator function, as that would've allowed
counting all sorts of other exceptions, but that would've dirtied design.

Finally, I just wasn't sure what kind of exception to expect here - are these always unhandled ones?
Or could they also be other `5xx` HTTP exceptions? Do 4xx exceptions arrive here at all (only confident that `404`s do not show up here)?

Perhaps you have some insight on this.

rycus86 commented on prometheus_flask_exporter/__init__.py:
```diff
@@ -353,21 +354,28 @@ def export_defaults(self, buckets=None, group_by='path',
         labels = self._get_combined_labels(None)
 
-        histogram = Histogram(
+        request_duration_metric = Histogram(
```
Thanks for renaming these! 👍

rycus86 commented on prometheus_flask_exporter/__init__.py:
```diff
@@ -353,21 +354,28 @@ def export_defaults(self, buckets=None, group_by='path',
+        request_exceptions_metric = Counter(
+            '%shttp_request_exceptions_total' % prefix,
+            'Total number of HTTP requests',
+            ('method', 'path', 'status') + labels.keys(),
```
Can we please leave the path off for now to be consistent with the other default metrics?
Perhaps another change could introduce a new parameter to also include a path (or endpoint) in the default labels, but I wouldn't add it here unconditionally.
I remember there were requests from people who have a large range of paths handled by a relatively small number of endpoints.

rycus86 commented on prometheus_flask_exporter/__init__.py:
```diff
@@ -413,24 +421,32 @@ def teardown_request(exception=None):
 
             response = make_response('Exception: %s' % exception, 500)
 
+            if callable(duration_group):
```
Curious why this moved here?

deepbrook replied:
Ah, right - I forgot to update the labels (see 'path' in your above comment) - instead, the exceptions metrics should also use `group`. Since we'll always be using it, no need to have it in the conditional below. I'll update the labels

rycus86 commented on prometheus_flask_exporter/__init__.py:
```diff
@@ -413,24 +421,32 @@ def teardown_request(exception=None):
-                    group = getattr(request, duration_group)
-
-                histogram_labels = {
+                request_duration_labels = {
```
Was this supposed to be this metric? It has hardcoded `500` for status?

deepbrook replied:
Oh crap - sorry, that was a copy-pasta error, same below

rycus86 replied:
No worries! Now that you found a better way to find the status code, should we use it for this one too instead of the hardcoded 500?

rycus86 commented on prometheus_flask_exporter/__init__.py:
```diff
@@ -413,24 +421,32 @@ def teardown_request(exception=None):
+
+            request_exceptions_labels = {
+                'method': request.method,
+                'status': 500,
```
Not sure if this should be hardcoded or inferred from the kind of exception.
Also, if status is always 500, why have it as a label at all?

deepbrook replied:
Ah, yes - I think I got confused with the hard-coded 500 during the histogram update on master. let me clean this up, I'll ping you once that's done.

rycus86 commented on tests/test_defaults.py:
```diff
@@ -109,6 +114,28 @@ def test():
+
+        self.assertMetric(
+            'flask_http_request_exceptions', '0.0',
+            ('method', 'GET'), ('path', '/error'), ('status', 200)
```
Isn't this supposed to be 1 here?

deepbrook replied:
yupp!

rycus86 replied:
I'm a bit confused about this test. It doesn't export the defaults but we do need them (you added a new default metric).
Also, this still needs to say `1.0` because we have already requested the error endpoint.

deepbrook commented:
@rycus86 have another look now. 

I've adjusted the labels to conform with the existing default metrics' labels.
I also cleaned up the copy-pasta errors I had, and the status code for the exception tracker in `teardown_request` should now do its best to divine a status_code from the exception, as long as it's a werkzeug `HTTPException`. In all other cases it just marks it as a `500`. If you have a better approach, let me know!

rycus86 commented on prometheus_flask_exporter/__init__.py:
```diff
@@ -413,24 +421,42 @@ def teardown_request(exception=None):
+            try:
+                response = exception.get_response()
+                status_code = _to_status_code(response.status_code)
+            except AttributeError:
```
Can we please catch all exceptions here to avoid accidentally masking the original exception with our problems?
```suggestion
            except:
```

rycus86 commented:
Looks like the Travis integration is borked again. 😔
This is the build for your latest change: https://travis-ci.org/github/rycus86/prometheus_flask_exporter/builds/717627876
I'll try to see if I can get it working and posting result here.

rycus86 commented:
OK, I think it will show up here now under the checks.

deepbrook commented on tests/test_defaults.py:
```diff
@@ -109,6 +114,28 @@ def test():
+            'flask_http_request_exceptions_total', '1.0',
+            ('method', 'GET'), ('path', '/error'), ('status', 501)
+        )
+
```
These asserts somehow do not want to pass - am I missing something here, @rycus86 ?
As they fail across 4 tests, perhaps I didn't declare/init the related metric proplery?

rycus86 replied:
Drop the `export_defaults=False` from the setup on line 118 - this is a new default metric, so you do want the defaults exported.

deepbrook replied:
:facepalm: fixed that, but the regexp still won't match. :thinking: Judging from what I can see from the test output, there's no value for the metric present.

rycus86 replied:
Maybe try raising an exception instead of `abort`, maybe that's handled internally by Flask, so perhaps it doesn't go to that exception path? (mostly guessing here)

deepbrook replied:
Tried that, but that appears to be not handled by the test client at all (had to catch it explicitly when calling `.get()` in the test, otherwise it would've errored).
hm.

rycus86 replied:
Ah yes, I forgot that the test client is a bit different. Aren't there any other tests that raise exceptions?

deepbrook replied:
There are, at least using `abort()`. It's still odd. The only explanation I can come up with is that the exception only arrives there if the error was not handlled by an error handler? But the code doesn't look that way (but following that around in the source of flask was tricky, so I might have missed something).

deepbrook commented on prometheus_flask_exporter/__init__.py:
```diff
@@ -413,24 +421,42 @@ def teardown_request(exception=None):
+            request_exceptions_labels = {
+                'method': request.method,
+                'status': status_code,
+                duration_group_name: group,
```
aach crud - this field isn't right, is it? The metrics shouldnt have a duration grouping to begin with.

rycus86 replied:
Ah yes, sorry I missed that. Yes, they probably shouldn't have it.

rycus86 commented:
Had a look at your change, and fixed the tests on https://github.com/rycus86/prometheus_flask_exporter/commits/help/deepbrook-add-requests-exceptions-counter-to-default-metrics
I squashed your commits into one, if it goes green I'll merge it an release a new version.
Thanks a lot @deepbrook !

rycus86 commented:
Also to mention, I removed the custom handling of `HTTPException`s as they came through the normal request end code path, not on the exceptional path.

rycus86 commented:
This is now merged in https://github.com/rycus86/prometheus_flask_exporter/commit/ab43abdc8387916c4746a4600c02d2098df8c959 and https://github.com/rycus86/prometheus_flask_exporter/commit/fdaf3f5de604cbd2d9e9ee6b99ada7273ebed4d0

Thanks a lot for this change!
The new metric will be available in version `0.16.0` that is currently releasing.
